utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -53,169 +86,317 @@ pub struct TryRuntimeCmd {
 		value_name = "METHOD",
 		possible_values = &WasmExecutionMethod::variants(),
 		case_insensitive = true,
-		default_value = "Interpreted"
+		default_value = "Compiled"
 	)]
 	pub wasm_method: WasmExecutionMethod,
 
-	/// The state to use to run the migration.
+	/// The number of 64KB pages to allocate for Wasm execution. Defaults to
+	/// sc_service::Configuration.default_heap_pages.
+	#[structopt(long)]
+	pub heap_pages: Option<u64>,
+}
+
+/// Various commands to try out against runtime state at a specific block.
+#[derive(Debug, Clone, structopt::StructOpt)]
+pub struct TryRuntimeCmd {
+	#[structopt(flatten)]
+	pub shared: SharedParams,
+
 	#[structopt(subcommand)]
-	pub state: State,
+	pub command: Command,
 }
 
-/// The state to use for a migration dry-run.
-#[derive(Debug, structopt::StructOpt)]
+/// The source of runtime state to try operations against.
+#[derive(Debug, Clone, structopt::StructOpt)]
 pub enum State {
-	/// Use a state snapshot as state to run the migration.
+	/// Use a state snapshot as the source of runtime state. NOTE: for the offchain-worker command this
+	/// is only partially supported at the moment and you must have a relevant archive node exposed on
+	/// localhost:9944 in order to query the block header.
+	// TODO https://github.com/paritytech/substrate/issues/9027
 	Snap {
 		snapshot_path: PathBuf,
 	},
 
-	/// Use a live chain to run the migration.
+	/// Use a live chain as the source of runtime state.
 	Live {
 		/// An optional state snapshot file to WRITE to. Not written if set to `None`.
 		#[structopt(short, long)]
 		snapshot_path: Option<PathBuf>,
 
 		/// The block hash at which to connect.
 		/// Will be latest finalized head if not provided.
-		#[structopt(short, long, multiple = false, parse(try_from_str = parse_hash))]
+		#[structopt(short, long, multiple = false, parse(try_from_str = parse::hash))]
 		block_at: Option<String>,
 
 		/// The modules to scrape. If empty, entire chain state will be scraped.
 		#[structopt(short, long, require_delimiter = true)]
 		modules: Option<Vec<String>>,
 
 		/// The url to connect to.
-		#[structopt(default_value = "ws://localhost:9944", parse(try_from_str = parse_url))]
+		#[structopt(default_value = "ws://localhost:9944", parse(try_from_str = parse::url))]
 		url: String,
 	},
 }
 
-fn parse_hash(block_number: &str) -> Result<String, String> {
-	let block_number = if block_number.starts_with("0x") {
-		&block_number[2..]
+async fn on_runtime_upgrade<B, ExecDispatch>(
```

Comment:
Life is easier when all generics have real names. Instead of `B`, how about `Block`?

Reply:
https://github.com/paritytech/substrate/pull/8966/commits/bb15b97e4dde17c65af32f701530b213aeac7a79